File: android/src/main/java/com/mapbox/examples/androidauto/AndroidAutoModule.java
//package com.mapbox.examples.androidauto;
//
//import android.content.Intent;
//import android.net.Uri;
//import android.util.Log;
//import android.os.Handler;
//import android.os.Looper;
//import androidx.activity.OnBackPressedCallback;
//import com.facebook.react.bridge.Callback;
//import com.facebook.react.bridge.ReactApplicationContext;
//import com.facebook.react.bridge.ReactContextBaseJavaModule;
//import com.facebook.react.bridge.ReactMethod;
//import com.facebook.react.bridge.ReadableMap;
//import com.facebook.react.bridge.WritableNativeMap;
//import com.facebook.react.module.annotations.ReactModule;
//import com.facebook.react.modules.core.DeviceEventManagerModule;
//import com.facebook.react.modules.debug.DevSettingsModule;
//import com.mapbox.api.directions.v5.DirectionsCriteria;
//import com.mapbox.api.directions.v5.MapboxDirections;
//import com.mapbox.api.directions.v5.models.DirectionsResponse;
//import com.mapbox.api.directions.v5.models.DirectionsRoute;
//import com.mapbox.api.directions.v5.models.RouteOptions;
//import com.mapbox.examples.androidauto.car.MainCarContext;
//import com.mapbox.examples.androidauto.car.MainCarScreen;
//import com.mapbox.examples.androidauto.car.MainScreenManager;
//import com.mapbox.examples.androidauto.car.preview.CarRoutePreviewScreen;
//import com.mapbox.examples.androidauto.car.preview.CarRouteRequest;
//import com.mapbox.examples.androidauto.car.preview.CarRouteRequestCallback;
//import com.mapbox.examples.androidauto.car.preview.RoutePreviewCarContext;
//import com.mapbox.examples.androidauto.car.search.PlaceRecord;
//import com.mapbox.examples.androidauto.car.search.SearchCarContext;
//import com.mapbox.geojson.Point;
//import com.mapbox.search.MapboxSearchSdk;
//import com.mapbox.search.ResponseInfo;
//import com.mapbox.search.ReverseGeoOptions;
//import com.mapbox.search.ReverseGeocodingSearchEngine;
//import com.mapbox.search.SearchCallback;
//import com.mapbox.search.SearchRequestTask;
//import com.mapbox.search.result.SearchResult;
//
//import androidx.annotation.NonNull;
//import androidx.car.app.CarContext;
//import androidx.car.app.CarToast;
//import androidx.car.app.Screen;
//import androidx.car.app.ScreenManager;
//import androidx.car.app.model.Template;
//
//import java.lang.reflect.Array;
//import java.util.Arrays;
//import java.util.List;
//import java.util.WeakHashMap;
//
//import kotlin.coroutines.CoroutineContext;
//import kotlinx.coroutines.BuildersKt;
//import kotlinx.coroutines.GlobalScope;
//import okhttp3.Route;
//import retrofit2.Call;
//import retrofit2.*;
//import retrofit2.Response;
//
//
//@ReactModule(name = AndroidAutoModule.MODULE_NAME)
//public class AndroidAutoModule extends ReactContextBaseJavaModule {
//
//  static final String MODULE_NAME = "CarModule";
//
//  private final Handler mHandler = new Handler(Looper.getMainLooper());
//  private static ReactApplicationContext mReactContext;
//  private MainCarContext mCarContext;
//  private MainScreenManager mMainScreenManager;
//  private ReactCarScreen mReactCarScreen;
//  private Screen mCurrentCarScreen;
//  private ScreenManager mScreenManager;
//  private String lastname;
//  private ReactCarRenderContext lastReactCarRenderContext;
//  private MapboxDirections client;
//
//
//  //@NonNull
//  //private SurfaceRenderer mSurfaceRenderer;
//
//  private WeakHashMap<String, Screen> carScreens;
//  private WeakHashMap<Screen, ReactCarRenderContext> reactCarRenderContextMap;
//
//  public String getName() {
//    return MODULE_NAME;
//  }
//
//  AndroidAutoModule(ReactApplicationContext context) {
//    super(context);
//    carScreens = new WeakHashMap();
//    reactCarRenderContextMap = new WeakHashMap();
//
//    mReactContext = context;
//  }
//
//  private ReverseGeocodingSearchEngine reverseGeocoding;
//  private SearchRequestTask searchRequestTask;
//
//  private final SearchCallback searchCallback = new SearchCallback() {
//
//    @Override
//    public void onResults(@NonNull List<? extends SearchResult> results, @NonNull ResponseInfo responseInfo) {
//      if (results.isEmpty()) {
//        Log.i("ReactAUTO", "No reverse geocoding results");
//      } else {
//        Log.i("ReactAUTO", "Reverse geocoding results: " + results);
//      }
//    }
//
//    @Override
//    public void onError(@NonNull Exception e) {
//      Log.i("ReactAUTO", "Reverse geocoding error", e);
//    }
//  };
//
//
//  private final CarRouteRequestCallback carRouteRequestCallback = new CarRouteRequestCallback() {
//    @Override
//    public void onRoutesReady(@NonNull PlaceRecord placeRecord, @NonNull List<? extends DirectionsRoute> routes) {
//      Log.d("ReactAUTO", "OnRoutesReady " + routes.toString());
//
//      mHandler.post(new Runnable() {
//        @Override
//        public void run() {
//          //mScreenManager.push(screen);
//          Screen screen = new CarRoutePreviewScreen(
//                  new RoutePreviewCarContext(mCarContext),
//                  placeRecord,
//                  (List<? extends DirectionsRoute>) routes.get(0)
//          );
//
//          reactCarRenderContextMap.remove(screen);
//          reactCarRenderContextMap.put(screen, lastReactCarRenderContext);
//          screen.setMarker(lastname);
//
//          carScreens.put(lastname, screen);
//          Log.d("ReactAUTO", carScreens.toString());
//          mCurrentCarScreen = screen;
//          Log.i("ReactAUTO", "pushScreen " + screen.toString());
//          mScreenManager.push(screen);
//        }
//      });
//    }
//
//    public void onUnknownCurrentLocation() {
//      Log.d("ReactAUTO", "onUnknownCurrentLocation");
//    }
//
//    public void onDestinationLocationUnknown() {
//      Log.d("ReactAUTO", "onDestinationLocationUnknown");
//    }
//
//    public void onNoRoutesFound() {
//      Log.d("ReactAUTO", "onNoRoutesFound");
//    }
//  };
//
//  @ReactMethod
//  public void invalidate(String name) {
//    Log.i("ReactAUTO", "invalidate AA Module");
//
//    Screen screen = getScreen(name);
//    mHandler.post(new Runnable() {
//      @Override
//      public void run() {
//        if (screen == mScreenManager.getTop()) {
//          screen.invalidate();
//        }
//      }
//    });
//  }
//
//  @ReactMethod
//  public void setTemplate(
//    String name,
//    ReadableMap renderMap,
//    Callback callback
//  ) {
//    ReactCarRenderContext reactCarRenderContext = new ReactCarRenderContext();
//    Screen screen = getScreen(name);
//    if (screen == null) {
//      screen = mCurrentCarScreen;
//      Log.i("ReactAUTO", "Screen " + name + " not found!");
//    }
//
//    reactCarRenderContext
//      .setEventCallback(callback)
//      .setScreenMarker(screen.getMarker());
//
//    Template template = parseTemplate(renderMap, reactCarRenderContext);
//    reactCarRenderContextMap.remove(screen);
//    reactCarRenderContextMap.put(screen, reactCarRenderContext);
//    Log.i("ReactAUTO", "Screen setTemplate" + template.toString());
//
//    ((MainCarScreen) screen).setTemplate(template);
//  }
//
//  @ReactMethod
//  public void pushScreen(
//    String name,
//    ReadableMap renderMap,
//    Callback callback
//  ) {
//    mHandler.post(new Runnable() {
//      @Override
//      public void run() {
//        ReactCarRenderContext reactCarRenderContext = new ReactCarRenderContext();
//        reactCarRenderContext.setEventCallback(callback).setScreenMarker(name);
//        
//        lastname = name;
//        lastReactCarRenderContext = reactCarRenderContext;
//
//         // TODO: parse template only if template provided
//        String type = renderMap.getString("type");
//        Screen screen;
//        // if type endswith template :
//        if(type.endsWith("template")){
//          Template template = parseTemplate(renderMap, reactCarRenderContext);
//          screen = mMainScreenManager.getNewScreen(false);
//          if(screen.getClass() == MainCarScreen.class) {
//            ((MainCarScreen) screen).setTemplate(template);
//          }
//        } else {
////          reverseGeocoding = MapboxSearchSdk.getReverseGeocodingSearchEngine();
////
////          final ReverseGeoOptions options = new ReverseGeoOptions.Builder(Point.fromLngLat(2.294434, 48.858349))
////                  .limit(1)
////                  .build();
////          searchRequestTask = reverseGeocoding.search(options, searchCallback);
//          Point origin = Point.fromLngLat(19.0, 47.0);
//          Point destination = Point.fromLngLat(19.2, 47.2);
//          RouteOptions routeOptions = RouteOptions.builder()
//                  .applyLanguageAndVoiceUnitOptions(this)
//// lists the coordinate pair i.e. origin and destination
//// If you want to specify waypoints you can pass list of points instead of null
//                  .coordinatesList(listOf(originPoint, destination))
//// set it to true if you want to receive alternate routes to your destination
//                  .alternatives(false)
//// provide the bearing for the origin of the request to ensure
//// that the returned route faces in the direction of the current user movement
//                  .bearingsList(
//                          listOf(
//                                  Bearing.builder()
//                                          .angle(originLocation.bearing.toDouble())
//                                          .degrees(45.0)
//                                          .build(),
//                                  null
//                          )
//                  )
//                  .build()
//          client = MapboxDirections.builder()
//                  .routeOptions()
//                  .destination(destination)
//                  .overview(DirectionsCriteria.OVERVIEW_FULL)
//                  .profile(DirectionsCriteria.PROFILE_DRIVING)
//                  .accessToken(getString(R.string.mapbox_access_token))
//                  .build();
//          client.enqueueCall(new retrofit2.Callback<DirectionsResponse>() {
//            @Override
//            public void onResponse(retrofit2.Call<DirectionsResponse> call, Response<DirectionsResponse> response) {
//              if (response.body() == null) {
//                Log.d("ReactAUTO", "No routes found, make sure you set the right user and access token.");
//                return;
//              } else if (response.body().routes().size() < 1) {
//                Log.d("ReactAUTO", "No routes found");
//
//                return;
//              }
//
//              DirectionsRoute currentRoute = response.body().routes().get(0);
//
//
//            }
//
//            @Override
//            public void onFailure(Call<DirectionsResponse> call, Throwable throwable) {
//              Log.d("ReactAUTO", "Error " + throwable.getMessage());
//            }
//          });
//
//
////          SearchCarContext searchCarContext = new SearchCarContext(mCarContext);
////          Point point = Point.fromLngLat(19.021870, 47.493226);
////          PlaceRecord placeRecord = new PlaceRecord(
////                  "some_id",
////                  "some_name",
////                  point,
////                  "some_description",
////                  Arrays.asList()
////          );
////          searchCarContext.getCarRouteRequest().request(placeRecord, carRouteRequestCallback);
//
////          ScreenParser screenParser = new ScreenParser();
////          screen = screenParser.getRoutePreviewScreen(mCarContext, renderMap, type);
//
//
//          return;
//
//
//        }
//        // else if endswith screen:
//
//        // else raise error
//
//
//
//        reactCarRenderContextMap.remove(screen);
//        reactCarRenderContextMap.put(screen, reactCarRenderContext);
//        screen.setMarker(name);
//
//        carScreens.put(name, screen);
//        Log.d("ReactAUTO", carScreens.toString());
//        mCurrentCarScreen = screen;
//        Log.i("ReactAUTO", "pushScreen " + screen.toString());
//        mScreenManager.push(screen);
//      }
//    });
//
//
//
//    mHandler.post(new Runnable() {
//      @Override
//      public void run() {
//        //mScreenManager.push(screen);
//      }
//    });
//  }
//
//  @ReactMethod
//  public void popScreen() {
//    mHandler.post(new Runnable() {
//      @Override
//      public void run() {
//        mScreenManager.pop();
//        removeScreen(mCurrentCarScreen);
//        mCurrentCarScreen = (MainCarScreen) mScreenManager.getTop();
//      }
//    });
//  }
//
//  @ReactMethod
//  public void mapNavigate(String address) {
//    mCarContext.getCarContext().startCarApp(
//      new Intent(CarContext.ACTION_NAVIGATE, Uri.parse("geo:0,0?q=" + address))
//    );
//  }
//
//  @ReactMethod
//  public void toast(String text, int duration) {
//    CarToast.makeText(mCarContext.getCarContext(), text, duration).show();
//  }
//
//  @ReactMethod
//  public void reload() {
//    DevSettingsModule devSettingsModule = mReactContext.getNativeModule(
//      DevSettingsModule.class
//    );
//    if (devSettingsModule != null) {
//      devSettingsModule.reload();
//    }
//  }
//
//  @ReactMethod
//  public void finishCarApp() {
//    mCarContext.getCarContext().finishCarApp();
//  }
//
//  @ReactMethod
//  public void setEventCallback(String name, Callback callback) {
//    MainCarScreen screen = (MainCarScreen) getScreen(name);
//
//    Log.d("ReactAUTO", "Set callback 1 for " + name);
//    if (screen == null) {
//      return;
//    }
//
//    Log.d("ReactAUTO", "Set callback 2 for " + name);
//    ReactCarRenderContext reactCarRenderContext = reactCarRenderContextMap.get(
//      screen
//    );
//
//    if (reactCarRenderContext == null) {
//      return;
//    }
//
//    Log.d("ReactAUTO", "Set callback 3 for " + name);
//    reactCarRenderContext.setEventCallback(callback);
//  }
//
//  public void setCarContext(MainCarContext carContext, MainScreenManager mainScreenManager) {
//    mCarContext = carContext;
//    mMainScreenManager = mainScreenManager;
//    mReactCarScreen = new ReactCarScreen(mCarContext);
//    mCurrentCarScreen = mainScreenManager.currentScreen(false);
//    mScreenManager = mCurrentCarScreen.getScreenManager();
//    carScreens.put("root", mCurrentCarScreen);
//
//    Log.d("ReactAUTO", carScreens.toString());
//
//    OnBackPressedCallback callback = new OnBackPressedCallback(true) {
//      @Override
//      public void handleOnBackPressed() {
//        Log.d("ReactAUTO", "Back button pressed");
//        mScreenManager.pop();
//        sendEvent("android_auto:back_button", new WritableNativeMap());
//      }
//    };
//
//    carContext.getCarContext().getOnBackPressedDispatcher().addCallback(callback);
//
//    sendEvent("android_auto:ready", new WritableNativeMap());
//  }
//
//  private Template parseTemplate(
//    ReadableMap renderMap,
//    ReactCarRenderContext reactCarRenderContext
//  ) {
//    TemplateParser templateParser = new TemplateParser(reactCarRenderContext);
//    return templateParser.parseTemplate(renderMap);
//  }
//
//  private Screen getScreen(String name) {
//    return carScreens.get(name);
//  }
//
//  private void removeScreen(Screen screen) {
//    WritableNativeMap params = new WritableNativeMap();
//    params.putString("screen", screen.getMarker());
//
//    sendEvent("android_auto:remove_screen", params);
//
//    carScreens.values().remove(screen);
//  }
//
//  private void sendEvent(String eventName, Object params) {
//    mReactContext
//      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
//      .emit(eventName, params);
//  }
//}

File: android/src/main/java/com/mapbox/examples/androidauto/AndroidAutoModule.kt
package com.mapbox.examples.androidauto

import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.activity.OnBackPressedCallback
import androidx.car.app.CarContext
import androidx.car.app.CarToast
import androidx.car.app.ScreenManager
import androidx.car.app.model.Template
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import com.facebook.react.modules.debug.DevSettingsModule
import com.mapbox.api.directions.v5.models.DirectionsRoute
import com.mapbox.examples.androidauto.car.MainCarContext
import com.mapbox.examples.androidauto.car.MainCarScreen
import com.mapbox.examples.androidauto.car.search.SearchCarContext
import com.mapbox.geojson.Point
import kotlinx.coroutines.InternalCoroutinesApi
import java.util.*
import java.util.WeakHashMap

import com.facebook.react.bridge.ReactApplicationContext
import com.mapbox.androidauto.*
import com.mapbox.api.directions.v5.DirectionsCriteria
import com.mapbox.api.directions.v5.models.RouteOptions
import com.mapbox.examples.androidauto.car.navigation.ActiveGuidanceScreen
import com.mapbox.examples.androidauto.car.navigation.CarActiveGuidanceCarContext
import com.mapbox.maps.extension.style.expressions.dsl.generated.step
import com.mapbox.navigation.base.extensions.applyDefaultNavigationOptions
import com.mapbox.navigation.base.route.RouterCallback
import com.mapbox.navigation.base.route.RouterFailure
import com.mapbox.navigation.base.route.RouterOrigin
import com.mapbox.turf.TurfMeasurement.distance


@ReactModule(name = AndroidAutoModule.MODULE_NAME)
class AndroidAutoModule(private val mReactContext: ReactApplicationContext) : ReactContextBaseJavaModule(mReactContext),
    DefaultLifecycleObserver {
    private val mHandler = Handler(Looper.getMainLooper())
    private var mCarContext: MainCarContext? = null
    private var mCurrentCarScreen: BaseCarScreen? = null
    private var mScreenManager: ScreenManager? = null
    private var screen: BaseCarScreen? = null
    private var mMainCarScreen: MainCarScreen? = null
    private var mActiveGuidanceScreen: ActiveGuidanceScreen? = null
    private var mReactScreen: BaseCarScreen? = null
    private var mStateConnected: Boolean = false
    private var currentRequestId: Long? = null

    private var carScreens: WeakHashMap<String, BaseCarScreen>
    private var reactCarRenderContextMap: WeakHashMap<BaseCarScreen, ReactCarRenderContext>


    private var onBackPressedCallback: () -> Unit = {
        Log.d("ReactAUTO", "Back: do nothing - set a proper action")
    }

    companion object {
        const val MODULE_NAME = "CarModule"
    }

    init {
        carScreens = WeakHashMap<String, BaseCarScreen>()
        reactCarRenderContextMap = WeakHashMap<BaseCarScreen, ReactCarRenderContext>()
        MapboxCarApp.updateCarAppState(ReactScreenState)
    }

    override fun getName(): String {
        return MODULE_NAME
    }

//    private val carAppStateObserver = Observer<CarAppState> { carAppState ->
//        val currentScreen = mCurrentCarScreen
//        val screenManager = mCarContext?.carContext?.getCarService(ScreenManager::class.java)
//        logAndroidAuto("--- MainScreenManager screen change ${currentScreen?.javaClass?.simpleName}")
//        if (screenManager?.top?.javaClass != currentScreen?.javaClass) {
//            screenManager?.push(currentScreen!!)
//        }
//    }

    fun getCurrentScreen(): BaseCarScreen {
        return mCurrentCarScreen ?: getScreenByState(ReactScreenState)
    }

    fun getScreenByState(carAppState: CarAppState, new: Boolean = false, root: Boolean = false): BaseCarScreen {
        logAndroidAuto("getScreenByState for state: $carAppState")
        //if(mCurrentScreen != null) return mCurrentScreen as MainCarScreen
        val screen = when (carAppState) {
            FreeDriveState, RoutePreviewState -> {
                if(new || mMainCarScreen == null){
                    mMainCarScreen = MainCarScreen(mCarContext!!)
                }
                mMainCarScreen!!

            }
//            ActiveGuidanceState, ArrivalState -> {
//                if(new || mActiveGuidanceScreen == null){
//                    mActiveGuidanceScreen = ActiveGuidanceScreen(
//                        CarActiveGuidanceCarContext(mCarContext!!),
//                        null,
//                        onBackPressedCallback,
//                        waypointPoints = waypointPoints
//                    )
//                }
//                mActiveGuidanceScreen!!
//
//            }
            ReactScreenState -> {
                if(new || mReactScreen == null){
                    mReactScreen = BaseCarScreen(mCarContext!!.carContext)
                }
                mReactScreen!!
            }
            else -> mMainCarScreen!!
        }
        //mCurrentCarScreen = screen
        if(root) screen.marker = "root"
        return screen
    }


    override fun onCreate(owner: LifecycleOwner) {
        logAndroidAuto("MainScreenManager onCreate")
        //MapboxCarApp.carAppState.observe(owner, carAppStateObserver)
    }

    override fun onDestroy(owner: LifecycleOwner) {
        logAndroidAuto("MainScreenManager onDestroy")
        //MapboxCarApp.carAppState.removeObserver(carAppStateObserver)
    }

    @ReactMethod
    fun invalidate(name: String) {
        Log.i("ReactAUTO", "invalidate AA Module")
        screen = getScreen(name)
        mHandler.post {
            if (screen === mScreenManager!!.top) {
                screen?.invalidate()
            }
        }
    }

    @ReactMethod
    fun setTemplate(
        name: String,
        renderMap: ReadableMap,
        callback: Callback?
    ) {
        val reactCarRenderContext = ReactCarRenderContext()
        screen = getScreen(name)
        if (screen == null) {
            screen = mCurrentCarScreen
            Log.i("ReactAUTO", "Screen $name not found!")
        }
        reactCarRenderContext
            .setEventCallback(callback).screenMarker = screen!!.marker
        val template = parseTemplate(renderMap, reactCarRenderContext) {
            Log.d("ReactAUTO", "Set onBackPressed")
            onBackPressedCallback = it
        }
        reactCarRenderContextMap.remove(screen)
        reactCarRenderContextMap[screen] = reactCarRenderContext
        Log.i("ReactAUTO", "Screen setTemplate$template")
        screen?.setTemplate(template)
    }

    @InternalCoroutinesApi
    @ReactMethod
    fun pushScreen(
        name: String?,
        renderMap: ReadableMap,
        callback: Callback?
    ) {
        mHandler.post {
            val reactCarRenderContext = ReactCarRenderContext()
            reactCarRenderContext.setEventCallback(callback).screenMarker = name

            // TODO: parse template only if template provided
            val type = renderMap.getString("type")
            // if type endswith template :
            if (type!!.endsWith("template")) {
                val template = parseTemplate(renderMap, reactCarRenderContext) {
                    Log.d("ReactAUTO", "Set onBackPressed")
                    onBackPressedCallback = it
                }
                screen = getScreenByState(ReactScreenState, true)
                screen?.setTemplate(template)
                reactCarRenderContextMap.remove(screen)
                reactCarRenderContextMap[screen] = reactCarRenderContext

                // else raise error
                screen?.marker = name
                carScreens[name] = screen
                Log.d("ReactAUTO", "name: $name")
                mCurrentCarScreen = screen
                Log.i("ReactAUTO", "pushScreen $screen")
                Log.d("ReactAUTO", carScreens.toString())
                mScreenManager!!.push(screen!!)

            } else {
                // else if endswith screen:
                var profile = DirectionsCriteria.PROFILE_DRIVING
                var waypointPoints = mutableListOf<Point>()
                //val manueverPoints = mutableMapOf<Point, String>()
                Log.d("ReactAUTO", "----------------------")
                Log.d("ReactAUTO", renderMap.toString())

                val metadata = renderMap.getArray("children")?.getMap(0)?.getMap("metadata")
                val waypoints = metadata?.getArray("waypoints")
                val destination = metadata?.getString("destination")
                if(metadata?.hasKey("offroad") == true && metadata.getBoolean("offroad") == true){
                    profile = DirectionsCriteria.PROFILE_CYCLING
                }
                for (i in 0 until waypoints?.size()!!) {
                    val waypoint = waypoints?.getMap(i)
                    // { NativeMap: {"data":{"type":"Feature","geometry":{"type":"Point","coordinates":[-123.714487,39.86334]},"properties":{"body":"Zero odometer at T junction. Turn left.","interval":"0.4","mileage":"0.0","direction":"arrow-left"}}} }
                    val data = waypoint?.getMap("data")
                    val geometry = data?.getMap("geometry")
                    //val properties = data?.getMap("properties")
                    //val body = properties?.getString("body")
                    val coordinates = geometry?.getArray("coordinates")
                    // Geometry: { NativeMap: {"type":"Point","coordinates":[-123.775287,39.577113]} }
                    //logAndroidAuto("waypoint $waypoint")
                    val point = Point.fromLngLat(
                        coordinates!!.getDouble(0),
                        coordinates.getDouble(1)
                    )
                    waypointPoints.add(point)
                    //manueverPoints[point] = body!!
                }

                Log.d("ReactAUTO", "Points: $waypointPoints")
                Log.d("ReactAUTO", "----------------------")
                val searchCarContext = SearchCarContext(mCarContext!!)
                // TODO: check distance to start point
                val location = MapboxCarApp.carAppServices.location().navigationLocationProvider.lastLocation
                val useFakePoints = false
                if(useFakePoints){
                    waypointPoints = listOf(
                        //Point.fromLngLat(location!!.longitude, location.latitude),
                        Point.fromLngLat(18.970608, 47.522207),
                        Point.fromLngLat(18.969998, 47.521167),
                        Point.fromLngLat(18.972462, 47.520275),
//                        Point.fromLngLat(18.974289, 47.518817),
//                        Point.fromLngLat(18.970734, 47.517787)
                    ) as MutableList<Point>
                }
                val maxDistanceAllowedInKm = 0.402336
                var alertTooFar = false
                if(
                    location == null ||
                    ( waypointPoints.size > 1 &&
                    distance(waypointPoints.get(0), Point.fromLngLat(location.longitude, location.latitude)) > maxDistanceAllowedInKm))
                {
                    alertTooFar = true
//                    screen = ErrorScreen(mCarContext!!.carContext)
//
//                    reactCarRenderContextMap.remove(screen)
//                    reactCarRenderContextMap[screen] = reactCarRenderContext
//                    screen?.marker = name
//                    carScreens[name] = screen
//                    Log.d("ReactAUTO", carScreens.toString())
//                    mCurrentCarScreen = screen
//                    Log.i("ReactAUTO", "pushScreen $screen")
//                    mScreenManager!!.push(screen!!)
                }
                // Navigation view
                Log.d("ReactAUTO", "Profile: $profile")
//                val waypointsForCalculation = if(location !== null && alertTooFar) {
//                    val tmp = waypointPoints.toMutableList()
//                    //tmp.add(0, Point.fromLngLat(location.longitude, location.latitude))
//                    tmp
//                } else waypointPoints
                currentRequestId?.let { searchCarContext.carRouteRequest.mapboxNavigation.cancelRouteRequest(it) }
                currentRequestId = searchCarContext.carRouteRequest.mapboxNavigation.requestRoutes(
                    RouteOptions.builder()
                        .applyDefaultNavigationOptions(profile)
                        .coordinatesList(waypointPoints)
                        .waypointIndicesList(listOf(0, waypointPoints.size-1))
                        .waypointNamesList(listOf("", destination))
                        .build(),
                    routesRequestCallback = object : RouterCallback {

                        override fun onRoutesReady(
                            routes: List<DirectionsRoute>,
                            routerOrigin: RouterOrigin
                        ) {
                            //Log.d("ReactAUTO", "onRoutesReady --- : $routes");

                            val carActiveGuidanceCarContext = CarActiveGuidanceCarContext(mCarContext!!)
                            val showNotification = try {
                                renderMap.getBoolean("notification")
                            } catch (e: NoSuchKeyException) {
                                false
                            }

                            //val stepBuilder = routes.get(0).legs()?.get(0)?.steps()?.get(0)?.toBuilder()
                            //stepBuilder?.maneuver(customemanouver)

//                                Log.d("ReactAUTO", "voice --- : \n${routes.get(0).legs()?.get(0)?.steps()?.get(0)?.voiceInstructions()}");
//                                Log.d("ReactAUTO", "bannerInstruction --- : \n${routes.get(0).legs()?.get(0)?.steps()?.get(0)?.bannerInstructions()}");
//                                Log.d("ReactAUTO", "intersections --- : \n${routes.get(0).legs()?.get(0)?.steps()?.get(0)?.intersections()}");
//                                Log.d("ReactAUTO", "maneuver --- : \n${routes.get(0).legs()?.get(0)?.steps()?.get(0)?.maneuver()}");

                            screen = ActiveGuidanceScreen(
                                carActiveGuidanceCarContext,
                                routes,
                                onBackPressedCallback,
                                showNotification,
                                alertTooFar,
                                waypointPoints,
                                //manueverPoints
                            )

                            reactCarRenderContextMap.remove(screen)
                            reactCarRenderContextMap[screen] = reactCarRenderContext
                            screen?.marker = name
                            carScreens[name] = screen
                            mCurrentCarScreen = screen
                            mScreenManager!!.push(screen!!)
                            carActiveGuidanceCarContext.mapboxNavigation.setRoutes(routes)
                        }

                        override fun onCanceled(
                            routeOptions: RouteOptions,
                            routerOrigin: RouterOrigin
                        ) {
                            Log.d("ReactAUTO", "RouteRequest onCanceled");

                        }

                        override fun onFailure(
                            reasons: List<RouterFailure>,
                            routeOptions: RouteOptions
                        ) {
                            Log.d("ReactAUTO", "RouteRequest onFailure");
                        }

                    }
                )




            }
        }
    }

    @ReactMethod
    fun popScreen() {
        Log.d("ReactAUTO", "popScreen");
        mHandler.post {
            mScreenManager!!.pop()
            removeScreen(mCurrentCarScreen)
            mCurrentCarScreen = mScreenManager!!.top as BaseCarScreen
        }
    }

    @ReactMethod
    fun mapNavigate(address: String) {
        mCarContext!!.carContext.startCarApp(
            Intent(CarContext.ACTION_NAVIGATE, Uri.parse("geo:0,0?q=$address"))
        )
    }

    @ReactMethod
    fun toast(text: String?, duration: Int) {
        CarToast.makeText(mCarContext!!.carContext, text!!, duration).show()
    }

    @ReactMethod
    fun reload() {
        val devSettingsModule = mReactContext.getNativeModule(
            DevSettingsModule::class.java
        )
        devSettingsModule?.reload()
    }

    @ReactMethod
    fun finishCarApp() {
        mCarContext!!.carContext.finishCarApp()
    }

    @ReactMethod
    fun setEventCallback(name: String, callback: Callback?) {
        screen = getScreen(name)
        Log.d("ReactAUTO", "Set callback 1 for $name")
        if (screen == null) {
            return
        }
        Log.d("ReactAUTO", "Set callback 2 for $name")
        val reactCarRenderContext = reactCarRenderContextMap[screen] ?: return
        Log.d("ReactAUTO", "Set callback 3 for $name")
        reactCarRenderContext.eventCallback = callback
    }

    @ReactMethod
    fun getState(){
        if(mStateConnected){
            sendEvent("android_auto:connect")
        } else {
            sendEvent("android_auto:disconnect")
        }
    }

    fun setCarContext(carContext: MainCarContext, startScreen: BaseCarScreen) {
        mCarContext = carContext
        mCurrentCarScreen = startScreen
        mScreenManager = mCurrentCarScreen?.screenManager
        carScreens["root"] = mCurrentCarScreen
        Log.d("ReactAUTO", carScreens.toString())
        val callback: OnBackPressedCallback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                Log.d("ReactAUTO", "Back button pressed")
                onBackPressedCallback()
                sendEvent("android_auto:back_button", WritableNativeMap())
            }
        }
        carContext.carContext.onBackPressedDispatcher.addCallback(callback)
        sendEvent("android_auto:ready", WritableNativeMap())
    }

    private fun parseTemplate(
        renderMap: ReadableMap,
        reactCarRenderContext: ReactCarRenderContext,
        setOnBackPressedCallback: (() -> Unit) -> Unit = {}
    ): Template {
        val templateParser = TemplateParser(mCarContext!!.carContext, reactCarRenderContext, setOnBackPressedCallback)
        return templateParser.parseTemplate(renderMap)
    }

    private fun getScreen(name: String): BaseCarScreen? {
        return carScreens[name]
    }

    private fun removeScreen(screen: BaseCarScreen?) {
        Log.d("ReactAUTO", "RemoveScreen2 $screen")

        val params = WritableNativeMap()
        params.putString("screen", screen!!.marker)
        sendEvent("android_auto:remove_screen", params)
        carScreens.values.remove(screen)
    }

    fun sendEvent(eventName: String, params: Any = WritableNativeMap()) {
        mReactContext
            .getJSModule(RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    fun connect(){
        mStateConnected = true
        getState()
    }

    fun disconnect(){
        mStateConnected = false
        getState()
    }


}